test/bandit_test.py

```diff
@@ -0,0 +1,321 @@
+            value_coefficients={"large": 2.0, "small": 1.0},
+        ),
+    ]
+    attributes = {}
+    expected_score = 0.2 + 0.3  # both are missing, use missing_value_coefficients
+    assert score_categorical_attributes(coefficients, attributes) == expected_score
+
+
+def test_score_categorical_attributes_empty_coefficients():
+    coefficients = []
+    attributes = {"color": "red", "size": "large"}
+    expected_score = 0.0  # no coefficients to apply
+    assert score_categorical_attributes(coefficients, attributes) == expected_score
+
+
+def test_score_categorical_attributes_negative_coefficients():
+    coefficients = [
+        BanditCategoricalAttributeCoefficient(
+            attribute_key="color",
+            missing_value_coefficient=0.2,
+            value_coefficients={"red": -1.0, "blue": -0.5},
+        ),
+        BanditCategoricalAttributeCoefficient(
+            attribute_key="size",
+            missing_value_coefficient=0.3,
+            value_coefficients={"large": -2.0, "small": -1.0},
+        ),
+    ]
+    attributes = {"color": "red", "size": "large"}
+    expected_score = -1.0 + -2.0
+    assert score_categorical_attributes(coefficients, attributes) == expected_score
+
+
+def test_score_categorical_attributes_mixed_coefficients():
+    coefficients = [
+        BanditCategoricalAttributeCoefficient(
+            attribute_key="color",
+            missing_value_coefficient=0.2,
+            value_coefficients={"red": 1.0, "blue": -0.5},
+        ),
+        BanditCategoricalAttributeCoefficient(
+            attribute_key="size",
+            missing_value_coefficient=0.3,
+            value_coefficients={"large": -2.0, "small": 1.0},
+        ),
+    ]
+    attributes = {"color": "blue", "size": "small"}
+    expected_score = -0.5 + 1.0
+    assert score_categorical_attributes(coefficients, attributes) == expected_score
+
+
+def test_weight_actions_single_action():
+    action_scores = [("action1", 1.0)]
+    gamma = 0.1
+    probability_floor = 0.1
+    expected_weights = [("action1", 1.0)]
+    assert (
+        bandit_evaluator.weight_actions(action_scores, gamma, probability_floor)
+        == expected_weights
+    )
+
+
+def test_weight_actions_multiple_actions():
+    action_scores = [("action1", 1.0), ("action2", 0.5)]
+    gamma = 0.1
+    probability_floor = 0.1
+    weights = bandit_evaluator.weight_actions(action_scores, gamma, probability_floor)
+    assert len(weights) == 2
+    assert any(action == "action1" and weight > 0.5 for action, weight in weights)
+    assert any(action == "action2" and weight <= 0.5 for action, weight in weights)
+
+
+def test_weight_actions_probability_floor():
+    action_scores = [("action1", 1.0), ("action2", 0.5), ("action3", 0.2)]
+    gamma = 0.1
+    probability_floor = 0.3
+    weights = bandit_evaluator.weight_actions(action_scores, gamma, probability_floor)
+    assert len(weights) == 3
+    for action, weight in weights:
+        assert weight >= 0.1
```

Comment:
Unless you disagree with the normalization

Reply:
Ah! I like the new normalization; I think I saw it after reviewing this test. I'd add a comment here explaining why it's 0.1, which otherwise looks like a typo (the value of gamma instead of probability_floor).